### PR TITLE
[CL Incentives] Implement proper accumulator checkpointing for positions `initOrUpdatePositionUptime`

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #4454

## What is the purpose of the change

Currently, when position records are added to an uptime accumulator, they checkpoint the global accumulator value. This over-distributes incentives, as each position should only be entitled to incentives that were accrued within their tick range.

## Brief Changelog

- Abstract all incentives-related position update logic into a helper for readability and proper testing:
   - If position doesn't exist (i.e. is new/being initialized), checkpoint `UptimeGrowthInside`
   - If position already exists, move up accum by `UptimeGrowthOutside` then update to new `UptimeGrowthInside`

## Testing and Verifying

- `initOrUpdatePositionUptime` is thoroughly tested in `incentive_test.go`
- Tests in `position_test.go` is updated to reflect new logic
 
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/concentrated-liquidity/incentives.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:368]: nit: we can call growthInside here, and growthOutside only needs to be called if the record does exist below.

File: x/concentrated-liquidity/lp.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp.go:128]: why?

**Review by czarcas7ic — APPROVED** [review]
This looks good on first pass! I would like to review it again, or get another pair of eyes on this before merge.

File: x/concentrated-liquidity/incentives.go
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:387]: We can move this outside of the loop. This would also be aligned with suggestion here: https://github.com/osmosis-labs/osmosis/pull/4455/files#r1120998514 to reduce the number of recomputations

File: x/concentrated-liquidity/incentives.go
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:345]: I know this was discussed before but can we separate init with update? Maybe in a post-completion refactor?

Currently, we are bundling two separate behaviors together, increasing the complexity and a likelihood of missing something small.

For ref, with fees, we init the position separately here: https://github.com/osmosis-labs/osmosis/blob/03f2a989a5b603ad9b34b90294bcbb7c7dae86cc/x/concentrated-liquidity/lp.go#L72-L74

and then doing the update here assuming that it is already initialized: https://github.com/osmosis-labs/osmosis/blob/03f2a989a5b603ad9b34b90294bcbb7c7dae86cc/x/concentrated-liquidity/lp.go#L225-L227

File: x/concentrated-liquidity/lp.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/lp.go:128]: #4456

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:345]: I believe that after #4468 is merged and positions are keyed by `joinTime` and `freezeDuration` (instead of just `frozenUntil`, we will be able to safely remove the ability to add to a frozen position, as the only case where it would be relevant becomes when someone tries to deposit to the same `freezeDuration` multiple times in the same block (see #4482).

Since this is not functionality we necessarily need to support (and actually disallowing it would serve as a great spam resistance vector), I think it will be appropriate to simplify this function to only handle initialization.

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:368]: Well, growthOutside returns the growth outside for all uptime accumulators so including it in the loop would lead to more calls right?

Also, after this change: #4482, this code won't require a call to growthOutside at all so this part will be removed anyway.

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go]: If you mean the position key computation then I agree – moved outside loop

**Review by stackman27 — APPROVED** [review]
LGTM, will refactor the functions once #4468 gets merged

**Pull request merged**
